File: python/block_parity.py
def block_encode(message):
    """Кодирование сообщения блочным кодом с проверкой на четность"""
    # Разбиваем сообщение на блоки по 8 бит
    blocks = [message[i:i + 8] for i in range(0, len(message), 8)]

    encoded_blocks = []
    for block in blocks:
        # Вычисляем бит четности
        parity_bit = str(block.count('1') % 2)
        # Добавляем бит четности в начало блока
        encoded_block = parity_bit + block
        encoded_blocks.append(encoded_block)

    # Объединяем закодированные блоки в одну строку
    encoded_message = ''.join(encoded_blocks)
    return encoded_message


def block_decode(encoded_message):
    """Декодирование сообщения, закодированного блочным кодом с проверкой на четность"""
    decoded_blocks = []
    for i in range(0, len(encoded_message), 9):
        # Извлекаем бит четности и блок данных из закодированного блока
        parity_bit, block = encoded_message[i], encoded_message[i + 1:i + 9]
        # Проверяем, правильно ли закодирован блок
        if str(block.count('1') % 2) == parity_bit:
            # Если блок закодирован правильно, удаляем бит четности и сохраняем блок данных
            decoded_blocks.append(block[0:])
        else:
            # Если блок закодирован неправильно, генерируем ошибку и завершаем декодирование
            raise ValueError("Ошибка декодирования: блок закодирован неправильно")

    # Объединяем декодированные блоки в одну строку
    decoded_message = ''.join(decoded_blocks)
    return decoded_message


# etel = encode("0000000110000001")
# decode(etel)

File: lua_python_complex/block_parity.lua
-- block_parity.lua
local BlockParity = {}

--- Кодирование сообщения блочным кодом с проверкой на четность
--- @param message string -- Сообщение для кодирования
--- @return string -- Закодированное сообщение
function BlockParity.block_encode(message)
    print("in block encode")
    print("message", message)
    local encoded_blocks = {}
    -- Разбиваем сообщение на блоки по 8 бит
    for i = 1, #message, 8 do
        local block = message:sub(i, i + 7)
        -- Вычисляем бит четности
        local parity_bit = #block:gsub("0", "") % 2
        -- Добавляем бит четности в начало блока
        table.insert(encoded_blocks, tostring(parity_bit) .. block)
    end
    -- Объединяем закодированные блоки в одну строку
    return table.concat(encoded_blocks)
end

--- Декодирование сообщения блочным кодом с проверкой на четность
--- @param encoded_message string -- Закодированное сообщение
--- @return string -- Декодированное сообщение
function BlockParity.block_decode(encoded_message)
    local decoded_blocks = {}
    for i = 1, #encoded_message, 9 do
        local block = encoded_message:sub(i, i + 8)
        -- извлекаем бит четности
        local parity_bit = tonumber(block:sub(1, 1))
        -- извлекаем данные блока
        local data_block = block:sub(2)
        -- Проверяем, правильно ли закодирован блок
        if (#data_block:gsub("0", "") % 2) == parity_bit then
            -- Если блок закодирован правильно, добавляем его к результату
            table.insert(decoded_blocks, data_block)
        else
            -- Если блок закодирован неправильно, выбрасываем ошибку
            error("Ошибка декодирования: блок закодирован неправильно")
        end
    end
    return table.concat(decoded_blocks)
end

return BlockParity
